content/Dev/python_good_practices_2020.rst

```diff
@@ -0,0 +1,371 @@
+
+
+The CircleCI configuration file is as simple as:
+
+.. code-block:: yaml
+
+    version: 2
+    jobs:
+      test:
+        docker:
+          - image: circleci/python:3.8
+        steps:
+          - checkout
+
+          - run:
+              name: Code lint
+              command: make lint
+
+          - run:
+              name: Test
+              command: make tests
+
+
+Code quality
+------------
+
+Running `black <https://black.readthedocs.io>`_ to format the code is now a no-brainer. We added `isort <https://github.com/timothycrosley/isort>`_ to sort and organize imports automatically too.
+
+The working combination in one ``Makefile`` target is:
+
+.. code-block:: make
+
+    format: $(INSTALL_STAMP)
+        $(VENV)/bin/isort --line-width=88 --lines-after-imports=2 -rc $(SOURCE) --virtual-env=$(VENV)
+        $(VENV)/bin/black $(SOURCE)
+
+Again, to avoid having an extra configuration file for *isort* we'll use CLI arguments :)
+
+We'll want to verify linting on our CI, so we have this ``lint`` target, that also runs `flake8 <https://pypi.org/project/flake8/>`_ to detect unused imports or variables, and run `mypy <http://mypy-lang.org/>`_ for type checking.
+
+.. code-block:: make
+
+    lint: $(INSTALL_STAMP)
+        $(VENV)/bin/isort --line-width=88 --check-only --lines-after-imports=2 -rc $(SOURCE) --virtual-env=$(VENV)
+        $(VENV)/bin/black --check $(SOURCE) --diff
+        $(VENV)/bin/flake8 $(SOURCE) --ignore=W503,E501
+        $(VENV)/bin/mypy $(SOURCE) --ignore-missing-imports
+
+By the way, using type checking is pretty straightforward and enjoyable :)
+
+.. code-block:: python
+
+    from typing import Any, Dict, List, Optional
+
+    def process(params: Optional[Dict[str, Any]] = None) -> List[str]:
+        return params.keys() if params else []
+
+Some plugins to garantee the quality of your contributions exist for your favorite editor. And a commit-hook can also do the job:
+
+.. code-block:: bash
+
+    echo "make format" > .git/hooks/pre-commit
+
+Check out `pre-commit <https://pre-commit.com>`_ or Rehan's `therapist <https://github.com/rehandalal/therapist>`_ for advanced commit hooks.
+
+Note that there are more linting tools out there:
+
+- `flake8-docstrings <https://pypi.org/project/flake8-docstrings/>`_ or `darglint <https://github.com/terrencepreilly/darglint>`_ to validate your docstrings
+- `wemake-python-styleguide <https://github.com/wemake-services/wemake-python-styleguide#what-we-are-about>`_ for a very strict Python linter
+- `bandit <https://bandit.readthedocs.io/en/latest/>`_ to find common security issues
+
+
+Tests
+-----
+
+There's almost no debate about `pytest <https://pytest.readthedocs.io>`_. To me, the most appealing feature is the `fixtures decorator <https://docs.pytest.org/en/latest/fixture.html>`_, to maintain your tests `DRY <https://en.wikipedia.org/wiki/Don%27t_repeat_yourself>`_. It allows to obtain dependency injection, object factories, connection setup, config changes...
```

Comment:
I would say `to keep your tests DRY`. Also, I wouldn't normally flag this, but I would not say `It allows [infinitive]` but rather `It allows you [infinitive]` or `It allows [gerund]`, and I wouldn't use the verb `obtain` for dependency injection -- maybe `use` or `perform` instead. Maybe rephrase as `It enables you to use dependency injection, object factories, connection setup, config changes...`